fly-apps/dockerfile-laravel
# Use upstream base image

**Discussion**: [See discourse post.](https://flyio.discourse.team/t/how-to-approach-securtiy-patches-for-fly-apps-laravel-docker/6484/11)

**TLDR**: we're porting the [fly-apps/laravel-docker](https://github.com/fly-apps/laravel-docker) functionality over to this package, so that this package may create a Dockerfile that does not use the custom image built by that repository, but uses an upstream image instead.

______________________________________

**WHAT AND WHY**
Revise Dockerfile template to create a Dockerfile that uses an upstream base image. There are three reasons for this:
1. **To remove the scope of having to maintain the custom image**( [fideloper/fly-laravel](https://hub.docker.com/r/fideloper/fly-laravel/tags) ) built from the [fly-apps/laravel-docker](https://github.com/fly-apps/laravel-docker) repository for SOC2 compliance, which we currently cannot fully commit to.
2. **Consistency**: The generated Dockerfiles for other frameworks like Rails, Elixir, Javascript, all use upstream images as base. For consistency, we should also follow for the Dockerfile generated for Laravel apps.
3. Finally, this move will result in depositing server config files to the users' project directory where they can **visibly and readily find them, providing them full control** over these files.

____________________________
**HOW**
Well super easy! [fly-apps/laravel-docker](https://github.com/fly-apps/laravel-docker) actually uses an upstream image as base, so we can simply merge the Dockerfile's logic it creates with our existing Dockerfile's logic, along with the necessary config files used by that Dockerfile to dockerfile-laravel. 

__________________________
**How To Review**
Ok! There're a lot of files right? lots. 
The easiest way to review this PR would be to see if it works for a local Laravel app one might have:
1. Get the PR locally
2. Run `pwd`---this should get the directory where the local dockerfile-laravel can be retrieved from
2. CD into a Laravel project
3. run: `php <path\to\dockerfile-laravel-project>\dockerfile-laravel generate --upstream-base-image`
4. Once done! Please update the .dockerignore to remove .env from files to ignore, since we need it for a local image
5. Please build the Dockerfile with `docker build -t my_sample_app .`
6. Run the image as container: `docker run -p 8080:8080 my_sample_app`

Hopefully everything works out well!

Lastly if you really wanna be thorough, here are the important files that got them changes:

**Primary Logic**
1. [app/Services/Scanner.php](https://github.com/fly-apps/dockerfile-laravel/pull/44/files#diff-fe405c70bcb7be8540ca3d17c2f8da250d5e041989309b517f3da3dbc8983da0) -  adds the necessary files into templates to render
2. [resources/views/dockerfile.blade.php](https://github.com/fly-apps/dockerfile-laravel/pull/44/files#diff-dc88aeb7892f11a327e6ff1c25d9f65a8a3390751b68c35dd009521dde8792f7) - revised to use laravel-docker's Dockerfile statements, and merge it with previous Dockerfile statements. This results in a Dockerfile that makes use of upstream image 

**Additional Templates**
There're a lot of these.
`resources/views/fly` -  should contain subdirectories and files that are templates to be rendered into the user's .fly directory

**The Tests**
These are not quite essential. but. _essential, too_ at the same time! Feel free to check one by one.
